### DISCUSSION TITLE
Self healing audio streams with codecs?

### DISCUSSION THREAD

**Discussion by joba-1** [pschatzmann/arduino-audio-tools#1069] [Q&A] [answered]
Hi,

I'd like to use a codec to reduce bandwidth to transfer audio over RS485 (currently bandwidth is ok for a single one way stream, but I want to send in both directions without collisions on the half duplex line where it will get tight).

As a first step I tried ADPCM (docs sounded promising) and it basically works (much easier than I thought), but...

Serial communication is not error free, so sooner or later the encoded stream is corrupted.
The decoder then spits out errors and never recovers.

Are there other light weight and still efficient (low cpu cost, high compression, lossy) codecs that would self heal after a short period after the transmission error?
Or alternatively: how can I sneak in between these two, detect an error and restart as fast as possible:

```c++
EncodedAudioStream dec(&sink, new ADPCMDecoder(AV_CODEC_ID_ADPCM_IMA_WAV));
StreamCopy copier(dec, Serial1, 1024); 
```

Currently I detect how many samples per second I get in the sink and restart the decoder if it is zero. That works, but not good enough...

**Comment by pschatzmann** [comment]
I think the issue is that if you loose some bytes the decoding stops to work because it uses the number of bytes to determine the frame start.

I would expect that it helps to pack the encoded data into some container. 

Recently I tried to add some [forward error correction](https://pschatzmann.github.io/arduino-audio-tools/group__fec.html) that might also help. However I did not have the time to test this yet...

**Comment by joba-1** [comment]
> I think the issue is that if you loose some bytes the decoding stops to work because it uses the number of bytes to determine the frame start.
> 
> I would expect that it helps to pack the encoded data into some container.

Cool, this shoud work. I'll try. 

I hoped the decoder could detect that a frame start is not like it should be (e.g. by not seeing some magic number in a header) and reinits like at the start, because if I reset the receiver the decoder (usually) finds a valid entry of the incoming stream.

Btw. what I described as restarting the decoder is effectively doing a full ESP32 reset, because this always crashes:

```
void loop() {
  copier.copy();
  if( decoder_error ) {
    LOGE("Decoder restart")
    decoder_error = false;
    dec.clearWriteError();
    dec.decoder().end();  // or dec.end()
    dec.decoder().begin();  // or dec.begin()
  }
}
```
shouldn't that work?


In the meantime I tried some other codecs (I found your blog):
* SBC works, but hangs the decoding ESP on line errors.
* APTX just produces cracks unrelated to the audio I send (always not just on line errors)
* LC3 works until line errors occur, but seems to not notice them, so I dont know how to react on it starting to produce crackling similar to APTX.

**Comment by pschatzmann** [comment]
You could check the result of copier.copy(); to determine if you were getting any data. But if you don't have any reliable way to determine the start of the frame, just restarting will not help.

I am curious if e.g my BinaryContainerEncoder/BinaryContainerDecoder is resolving this issue...

**Reply by joba-1** [reply]
Have to dig into using the copy() result. It looks like there is no explicit error return code, just size 0. And this could also be a valid result in a fast loop? At least I would get rid of the global decoder_error flag that I set in my custom stream class.

I wrapped the decoder and encoder in your containers (wow, your interfaces are all so easy to use!) and it works, but is more fragile:
I must start sender and receiver several times. Most of the time I get 
lots of `[W] ContainerBinary.h : 331 - data ignored` on the receiver immediately. But even if it works for a while (seconds) eventually these messages start and I only get silence from then on.
I probably need to do some error checking...

**Comment by pschatzmann** [comment]
I havn't done some heavy testing on my ContainerBinary: so there might still be some bugs in it...

**Reply by joba-1** [reply]
nobody is perfect :)

And maybe I violate "We expect that a single write() is providing full frames." from the doxygen?
I copier.copy() from Serial1 with default buffersize (1024) to the decoder/container. Not sure how/if this copies a frame in one go.

I implemented the "no data" detection from the copy() result and that works as before with my global flag if I do not use the container. Including the same problems: decoder.end() leads to ESP reset. Also with the container if the sender is not sending at all:

```c++
void loop() {
  static const uint32_t Timeout = 1000;
  static uint32_t last_copy = 0;
  uint32_t now = millis();
  if( copier.copy() ) {
    last_copy = now;
  }
  else if( now - last_copy > Timeout ) {
    LOGE("Decoder restart")
    dec.clearWriteError();
    dec.decoder().end();   // src/main.cpp:167 see stack trace below
    dec.decoder().begin();
    last_copy = now;
  }
}

```
```
[E] main.cpp : 165 - Decoder restart
[I] CodecADPCM.h : 47 - virtual void audio_tools::ADPCMDecoder::end()
Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.

Core  1 register dump:
PC      : 0x400d7a18  PS      : 0x00060a30  A0      : 0x800d7a38  A1      : 0x3ffc7ab0  
A2      : 0x00000000  A3      : 0x3f4001a0  A4      : 0x00000002  A5      : 0x0000ff00  
A6      : 0x00ff0000  A7      : 0xff000000  A8      : 0x800d7a18  A9      : 0x3ffc7aa0  
A10     : 0x00000001  A11     : 0x3f4001a1  A12     : 0x000000ff  A13     : 0x0000ff00  
A14     : 0x00ff0000  A15     : 0xff000000  SAR     : 0x0000000a  EXCCAUSE: 0x0000001c  
EXCVADDR: 0x00000000  LBEG    : 0x40087601  LEND    : 0x40087611  LCOUNT  : 0xffffffff  


Backtrace: 0x400d7a15:0x3ffc7ab0 0x400d7a35:0x3ffc7ad0 0x400d274f:0x3ffc7af0 0x400d5570:0x3ffc7b10 0x4011ea72:0x3ffc7b30 0x400d244d:0x3ffc7b50 0x4011ea7d:0x3ffc7b70 0x400d244d:0x3ffc7b90 0x400d4c56:0x3ffc7bb0 0x400d886d:0x3ffc7bd0

  #0  0x400d7a15:0x3ffc7ab0 in Print::write(char const*) at /home/joachim/.platformio/packages/framework-arduinoespressif32/cores/esp32/Print.h:67
  #1  0x400d7a35:0x3ffc7ad0 in Print::print(char const*) at /home/joachim/.platformio/packages/framework-arduinoespressif32/cores/esp32/Print.cpp:84
  #2  0x400d274f:0x3ffc7af0 in audio_tools::AudioLogger::printPrefix(char const*, int, audio_tools::AudioLogger::LogLevel) const at .pio/libdeps/mhetesp32minikit_ser/audio-tools/src/AudioTools/AudioLogger.h:123
  #3  0x400d5570:0x3ffc7b10 in audio_tools::AudioLogger::prefix(char const*, int, audio_tools::AudioLogger::LogLevel) at .pio/libdeps/mhetesp32minikit_ser/audio-tools/src/AudioTools/AudioLogger.h:52 (discriminator 1)
      (inlined by) audio_tools::ADPCMDecoder::end() at .pio/libdeps/mhetesp32minikit_ser/audio-tools/src/AudioCodecs/CodecADPCM.h:47 (discriminator 1)
  #4  0x4011ea72:0x3ffc7b30 in audio_tools::ContainerTargetPrint::end() at .pio/libdeps/mhetesp32minikit_ser/audio-tools/src/AudioCodecs/AudioEncoded.h:630 (discriminator 1)
  #5  0x400d244d:0x3ffc7b50 in audio_tools::BinaryContainerDecoder::end() at .pio/libdeps/mhetesp32minikit_ser/audio-tools/src/AudioCodecs/ContainerBinary.h:227
  #6  0x4011ea7d:0x3ffc7b70 in audio_tools::ContainerTargetPrint::end() at .pio/libdeps/mhetesp32minikit_ser/audio-tools/src/AudioCodecs/AudioEncoded.h:631 (discriminator 1)
  #7  0x400d244d:0x3ffc7b90 in audio_tools::BinaryContainerDecoder::end() at .pio/libdeps/mhetesp32minikit_ser/audio-tools/src/AudioCodecs/ContainerBinary.h:227
  #8  0x400d4c56:0x3ffc7bb0 in loop() at src/main.cpp:167
  #9  0x400d886d:0x3ffc7bd0 in loopTask(void*) at /home/joachim/.platformio/packages/framework-arduinoespressif32/cores/esp32/main.cpp:50
```
If the "data ignored" appears, this timeout does not trigger anymore, i.e the loop() is no longer processed at all: endless loop somewhere (I guess in size_t BinaryContainerDecoder::write(const void *data, size_t len): ... while (open > 0) { result = processData(data8 + processed, open);...

**Comment by pschatzmann** [comment]
By the way, if you use the 8 bit codec you can still compress the audio by half and any lost byte will not disturb at all, it just might cause that left and right get switched if you play stereo.

**Reply by joba-1** [reply]
Sounds interesting. I get stereo from the mic. But since one channel is garbage anyways, I convert to one channel. So channel swap will not be a problem :)
Will try that next...

**Comment by pschatzmann** [comment]
Maybe the problem is somewhere else. 
What happens if the audio is sent faster than the receiver can play ? Is data getting lost or is the sender stalling until some receive buffers are available again ?

**Comment by joba-1** [comment]
There is no real sync. The full pipeline is:

## Sender
```c++
I2SStream i2s;  // INMP441 delivers 24 as 32bit
Convert024to16 cvt(i2s);  // convert 2ch 24bit to 1ch 16bit
BinaryContainerEncoder bcd(new ADPCMEncoder(AV_CODEC_ID_ADPCM_IMA_WAV));
EncodedAudioStream enc(&Serial1, &bcd);
StreamCopy copier(enc, cvt, 1024);  // data pump
```

## Receiver
```c++
I2SStream i2s;  // sink MAX98357A mono amp
Convert cvt(i2s);  // 1ch -> 2ch as required for the mono amp - go figure... :)
BinaryContainerDecoder bcd(new ADPCMDecoder(AV_CODEC_ID_ADPCM_IMA_WAV));
EncodedAudioStream dec(&cvt, &bcd);
StreamCopy copier(dec, Serial1, 1024); 
```
I guess the amp will just get garbage if no data is available in the pace it expects it. In fact that is what happened while I still used standard sample rates with 2 channels and no codec (the RS485 maxed out with that). This did not cause program errors.

**Reply by pschatzmann** [reply]
OK - I see: in this case it should be OK because I2S provides the data at the right rate...

I have this issue in my test code because I use a Sine Generator as source which provides the data much too fast....

**Comment by joba-1** [comment]
If you want to lookup any details: I just created these github repos:
* https://github.com/joba-1/esp32-i2s-mic
* https://github.com/joba-1/esp32-i2s-amp

**Comment by joba-1** [comment]
no instant success with L8: updated the repos.

It looks like throughput is very low (high load?). Usually, with ADPCM I see uncompressed 32kB/s as expected, now it is only about 5kB/s

```
[I] main.cpp : 103 - Amplitude 21672
[I] main.cpp : 101 - Written/2 5062 Bytes/s
[I] main.cpp : 102 - Read      5062 Bytes/s
[I] main.cpp : 103 - Amplitude 4902
[I] main.cpp : 101 - Written/2 4962 Bytes/s
[I] main.cpp : 102 - Read      4962 Bytes/s

```

sound is not recognizable, just loud cracks...

Will be offline for a bit. Thanks for your support so far!

**Reply by joba-1** [reply]
seen your notes in the PR. Will retry L8 and Container soon...

**Comment by pschatzmann** [comment, marked as answer]
The simplest solution is to send each audio sample as 1 byte. If some bytes get lost, this will be not audible at all and in terms of audio quality it is very hard to distinguish between 8 bit and 16 bits. Here is the [related example.](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-communication/serial/send-8bit-receive/send-8bit-receive.ino). To test this I was increasing the baud rate to a very high value...

For adpcm I have extended by binary container: it stores the audio info, metadata and audio in different record structures/segments. Each encoded frame is stored in a separate segment and a checksum is calculated to determine if the audio is still valid. The records start with a crlf so that the beginning of a segment can be found easily.  Here is the [related example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-communication/serial/send-adpcm_framed-receive/send-adpcm_framed-receive.ino)

**Reply by joba-1** [reply]
I just tried the adpcm example. Changed serial2 and i2s pins and used higher serial baud (460800): short sound on both sides, then breaks:

```
frame_size: 249E (34) I2S: i2s_driver_uninstall(2047): I2S port 0 has not installed
dex[0] = 13621
ERROR: step_indERROR: step_index[0] = -21315
ERROR: step_index[0] = 2459
ERROR: step_index[0] = 17714
ERROR: step_index[0] = 9012
ERROR: step_index[0] = -26607
ERROR: step_index[0] = -17189
...
```

And then the framed example (no sound):

```
frame_size: 121invalid number of samples in packet
invalid number of samples in packet
...
invalid number of samples in packet
[W] ContainerBinary.h : 343 - invalid checksum
[W] ContainerBinary.h : 343 - invalid checksum
invalid number of samples in packet
...
```

**Reply by joba-1** [reply]
and the 8bit one:

I can hear the sound permanently, not just a short time, but very choppy (like at each copy iteration, it changes with copy buffer size)

Test setup:

![image](https://github.com/pschatzmann/arduino-audio-tools/assets/32450554/f69d7634-d467-4d6c-b8e1-859a1ac1fc81)

**Reply by joba-1** [reply]
After a lot of fiddling I found out that by far most of the problems occur on the left ESP. It is a ESP32-D0WDQ6 while the right one is a ESP32-D0WD-V3. Maybe that matters? I'll replace it...

**Reply by joba-1** [reply]
...and success - I can hear sine waves or very clear i2s-mic sound (without any choppyness as before) on the other esps respectively! Very happy right now :)

so two ESP32-D0WD-V3 and most problems are gone - I could increase sample rate from 8k to 16k (not tried more yet, won't need it).

What is left is that I often need more than one boot, before the ESPs stop just spitting out `ERROR: step_index[0] = <some int16>`  instead of sound and volume is quite low (earlier I had to watch out for feedback loops from the mics, not anymore). After that I'll try RS485 half duplex with packets.
Ah, and those messages during setup() are a bit alarming:
```
E (27) I2S: i2s_driver_uninstall(2047): I2S port 0 has not installed
E (30) I2S: register I2S object to platform failed
[E] I2SESP32.h : 186 - begin - i2s_driver_install
E (43) I2S: i2s_driver_uninstall(2047): I2S port 1 has not installed
```

So lesson learned: ESP32-D0WDQ6 (Rev 0) is no good for I2S...

**Comment by joba-1** [comment]
For reference, this is the code that works:

```c++
/**
 * Derived from
 * 
 * @file send-adpcm-receive.ino
 * @author Phil Schatzmann
 * @brief Sending and receiving audio via Serial. You need to connect the RX pin
 * with the TX pin!
 * 
 * We send encoded ADPCM audio over the serial wire: The higher the transmission rate
 * the higher the risk of data loss!
 *
 * @version 0.1
 * @date 2023-11-25
 *
 * @copyright Copyright (c) 2022
 */

#include "AudioTools.h"
#include "AudioCodecs/CodecADPCM.h" // https://github.com/pschatzmann/adpcm
//#include "AudioLibs/AudioKit.h"

AudioInfo info(16000, 1, 16);
I2SStream out; // or AnalogAudioStream, AudioKitStream etc
I2SStream in;
// SineWaveGenerator<int16_t> sineWave(32000);
// GeneratedSoundStream<int16_t> sineStream(sineWave);

auto &serial = Serial2;
ADPCMEncoder enc(AV_CODEC_ID_ADPCM_IMA_WAV);
ADPCMDecoder dec(AV_CODEC_ID_ADPCM_IMA_WAV);
EncodedAudioStream enc_stream(&serial, &enc);
EncodedAudioStream dec_stream(&out, &dec);
// Throttle throttle(enc_stream);
static int frame_size = 256;
// StreamCopy copierOut(throttle, sineStream, frame_size);  // copies sound into Serial
StreamCopy copierOut(enc_stream, in, frame_size);  // copies mic into Serial
StreamCopy copierIn(dec_stream, serial, frame_size);  // copies sound from Serial


void inputTask( void * parameter ){
  Serial.printf("input() on core %d\n", xPortGetCoreID());
  while( true ) {
    // copy from serial
    copierIn.copy();
    delay(0);  // nop?
  }
}

void outputTask( void * parameter ){
  Serial.printf("output() on core %d\n", xPortGetCoreID());
  while( true ) {
    // copy to serial
    copierOut.copy();
    delay(0);  // nop?
  }
}

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  BaseType_t coreId = xPortGetCoreID();
  Serial.printf("setup() on core %d\n", coreId);
  Serial.printf("ESP model:  %s\n",    ESP.getChipModel());
  Serial.printf("ESP cores:  %u\n",    ESP.getChipCores());
  Serial.printf("ESP rev:    %u\n",    ESP.getChipRevision());
  Serial.printf("ESP freq:   %u\n",    ESP.getCpuFreqMHz());
  Serial.printf("ESP mac:    %08lx\n", ESP.getEfuseMac());
  Serial.printf("ESP fsize:  %u\n",    ESP.getFlashChipSize());
  Serial.printf("ESP fspeed: %u\n",    ESP.getFlashChipSpeed());
  Serial.printf("ESP fmode:  %u\n",    ESP.getFlashChipMode());
  Serial.printf("ESP heap:   %u\n",    ESP.getFreeHeap());
  Serial.printf("ESP psram:  %u\n",    ESP.getFreePsram());

  // Note the format for setting a serial port is as follows:
  // Serial.begin(baud-rate, protocol, RX pin, TX pin);
  Serial2.begin(921600, SERIAL_8N1, 18, 19);

  // sineWave.begin(info, N_B4*AMP);
  // throttle.begin(info);
  enc_stream.begin(info);
  dec_stream.begin(info);

  // PCM5102 SCK -> GND
  pinMode(22, OUTPUT);
  digitalWrite(22, LOW);

  // start I2Sin
  auto configIn = in.defaultConfig(RX_MODE);
  configIn.copyFrom(info);
  configIn.pin_data = 23;
  configIn.pin_bck = 5;
  configIn.pin_ws = 26;
  configIn.port_no = 0;
  in.begin(configIn);

  // start I2Sout
  auto configOut = out.defaultConfig(TX_MODE);
  configOut.copyFrom(info);
  configOut.pin_data = 17;
  configOut.pin_bck = 21;
  configOut.pin_ws = 16;
  configOut.port_no = 1;
  out.begin(configOut);

  // better visibility in logging
  copierOut.setLogName("out");
  copierIn.setLogName("in");

  xTaskCreatePinnedToCore(
    inputTask,        /* Task function. */
    "inputTask",      /* String with name of task. */
    10000,            /* Stack size in words. */
    NULL,             /* Parameter passed as input of the task */
    AMP,                /* configMAX_PRIORITIES - 1, Priority of the task. */
    NULL,             /* Task handle. */
    coreId ? 1 : 0);  /* same core id as main task */

  xTaskCreatePinnedToCore(
    outputTask,       /* Task function. */
    "outputTask",     /* String with name of task. */
    10000,            /* Stack size in words. */
    NULL,             /* Parameter passed as input of the task */
    3-AMP,                /* configMAX_PRIORITIES - 2, Priority of the task. */
    NULL,             /* Task handle. */
    coreId ? 0 : 1);  /* other core id than main task */
}

void loop() {
  static bool first = true;
  if( first ) {
    first = false;
    Serial.printf("loop() on core %d\n", xPortGetCoreID());
  }
  delay(100);
} 
```

with this platformio.ini

```
[platformio]
default_envs = mhetesp32minikit_2, mhetesp32minikit_1
; src_dir = send-adpcm_framed-receive
src_dir = send-adpcm-receive
; src_dir = send-8bit-receive

[program]
name = AdpcmRecv
version = 1.0
instance = 1

[env]
framework = arduino
platform = https://github.com/platformio/platform-espressif32.git
board = mhetesp32minikit
monitor_filters = esp32_exception_decoder
monitor_speed = 115200
lib_deps = 
    https://github.com/pschatzmann/arduino-audio-tools.git
    https://github.com/pschatzmann/adpcm.git
    # https://github.com/pschatzmann/arduino-libopus.git
    # can hang https://github.com/pschatzmann/arduino-libsbc.git
    # noise https://github.com/pschatzmann/arduino-libopenaptx.git
    # silent errors https://github.com/pschatzmann/arduino-liblc3.git
build_flags = 
    -Wall 
    -DPIO_FRAMEWORK_ARDUINO_ENABLE_EXCEPTIONS
    -DVERSION='"${program.version}"'
    -DPROGNAME='"${program.name}"'
    -DHOSTNAME='"${program.name}-${program.instance}"'
    -DBAUDRATE=${env.monitor_speed}
    -DCOPY_LOG_OFF

[env:mhetesp32minikit_1]
monitor_port = /dev/ttyACM0
upload_port = /dev/ttyACM0
build_flags =
    ${env.build_flags}
    -DAMP=2

[env:mhetesp32minikit_2]
monitor_port = /dev/ttyACM1
upload_port = /dev/ttyACM1
build_flags = 
    ${env.build_flags}
    -DAMP=1

```
